python/hail/type.py

```diff
@@ -1,12 +1,582 @@
+    """
+    Error thrown at mismatch between expected and supplied python types.
+
+    :param str message: Error message
+    """
+
+    def __init__(self, message):
+        self.msg = message
+        super(TypeCheckError).__init__(TypeCheckError)
+
+    def __str__(self):
+        return self.msg
+
 
 class Type(object):
-    """Type of values."""
-    
+    """
+    Hail type class used for annotations and expression language.
+
+    :param jtype: equivalent java type
+    """
+
     def __init__(self, jtype):
-        self.jtype = jtype
+        self._jtype = jtype
 
     def __repr__(self):
-        return self.jtype.toString()
+        return self._jtype.toString()
 
     def __str__(self):
-        return self.jtype.toPrettyString(False, False)
+        return self._jtype.toPrettyString(False, False)
+
+    def __eq__(self, other):
+        return self._jtype.equals(other._jtype)
+
+    @classmethod
+    def _from_java(cls, jtype):
+        # FIXME this is pretty hacky
+        class_name = jtype.getClass().getCanonicalName()
+        if class_name == 'is.hail.expr.TInt$':
+            return TInt()
+        elif class_name == 'is.hail.expr.TLong$':
+            return TLong()
+        elif class_name == 'is.hail.expr.TFloat$':
+            return TFloat()
+        elif class_name == 'is.hail.expr.TDouble$':
+            return TDouble()
+        elif class_name == 'is.hail.expr.TString$':
+            return TString()
+        elif class_name == 'is.hail.expr.TBoolean$':
+            return TBoolean()
+        elif class_name == 'is.hail.expr.TSample$':
+            return TString()
+        elif class_name == 'is.hail.expr.TArray':
+            return TArray._from_java(jtype)
+        elif class_name == 'is.hail.expr.TSet':
+            return TSet._from_java(jtype)
+        elif class_name == 'is.hail.expr.TDict':
+            return TDict._from_java(jtype)
+        elif class_name == 'is.hail.expr.TStruct':
+            return TStruct._from_java(jtype)
+        elif class_name == 'is.hail.expr.TVariant$':
+            return TVariant()
+        elif class_name == 'is.hail.expr.TAltAllele$':
+            return TAltAllele()
+        elif class_name == 'is.hail.expr.TLocus$':
+            return TLocus()
+        elif class_name == 'is.hail.expr.TInterval$':
+            return TInterval()
+        elif class_name == 'is.hail.expr.TGenotype$':
+            return TGenotype()
+        else:
+            raise TypeError("unknown type class: '%s'" % class_name)
+
+    def _jtype(self):
+        """
+        Returns the java type associated with this type.
+        """
+
+        raise NotImplemented('use a subclass of Type')
+
+    def _typecheck(self, annotation):
+        """
+        Raise an exception if the given annotation is not the appropriate type.
+
+        :param annotation: value to check
+        """
+        raise NotImplemented('use a subclass of Type')
+
+
+class TInt(Type):
+    """
+    Hail type corresponding to int
```

Comment:
You mean python int?  I don't think that's quite right.  The size of python ints are system dependent.  TInt is 32-bit integers.